OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+    return fiberLength;
+}
+double  VandenBogert2011Muscle::getActivation(SimTK::State& s) const
+{
+    double activ=getStateVariableValue(s, "activation");
+    return activ;
+}
+
+
+
+//==============================================================================
+// Muscle.h Interface
+//==============================================================================
+
+
+double  VandenBogert2011Muscle::computeActuation(const SimTK::State& s) const
+{
+    double activ=getActivation(s);
+    // For now we will start with static guess
+    // TODO: figure out how to use previous state as guess
+    SimTK::Vector ydotInitialGuess(2);
+    ydotInitialGuess[0]=0.0;
+    ydotInitialGuess[1]=0.0;
+    SimTK::Vec2 sdot=calcSolveMuscle(s, activ, ydotInitialGuess);
+
+    VandenBogert2011Muscle::ImplicitResidual Result= calcImplicitResidual(s,
+                                          sdot[0],
+                                          sdot[1],
+                                          activ,
+                                          0);
+
+    //TODO: Could probably add a check here that residuals =0
+    double actuation=Result.forceTendon;
+
+    return(actuation);
+}
+
+
+
+void VandenBogert2011Muscle::computeInitialFiberEquilibrium(SimTK::State& s)
+const
+{
+    // TODO:  this is the same code as computeFiberEquilibriumAtZeroVelocity
+    //Need to update for v~=0
+    //Let's start with the default activation and the fiberVelocity=0
+
+    computeFiberEquilibriumAtZeroVelocity(s);
+
+}
+
+
+
+void VandenBogert2011Muscle::
+computeFiberEquilibriumAtZeroVelocity(SimTK::State& s) const
+{
+    double activation = getDefaultActivation();
+    SimTK::Vec2 lenghAndForce=calcFiberStaticEquilbirum(getLength(s),
+                                                        activation);
+
+    double projFibLenNorm=lenghAndForce[0];
+    double tendonForce=lenghAndForce[1];
+
+    //TODO:  Millard Muscle handles non-convergence here and zero muscle length.
+    //     Need to consider adding.
+
+    setActuation(s, tendonForce);
+    double fiberLength =
+            projFibLenToFiberLength(projFibLenNorm,1)*getOptimalFiberLength();
+    setStateVariableValue(s, "fiber_length", fiberLength);
+    setStateVariableValue(s, "activation",  activation);
+
+}
+// I am cheating here.  Instead of using mli and mdi caches
+/*
+SimTK::Vec2 VandenBogert2011Muscle::getStateVariables(const SimTK::State& s) {
+    double fiberLength = getStateVariableValue(s, "fiber_length");
+    double activ = getStateVariableValue(s, "activation");
+    SimTK::Vec2 stateVariables = {fiberLength, activ};
+    return stateVariables;};
+
+*/
+
+
+//==============================================================================
+// MODELCOMPONENT INTERFACE REQUIREMENTS
+//==============================================================================
+
+void VandenBogert2011Muscle::extendConnectToModel(Model& model)
+{
+    Super::extendConnectToModel(model);
+}
```

Comment:
Not necessary to override this method if you are not going to do anything specific to this class. Millard2012 should not override it either.